token.go

```diff
@@ -0,0 +1,226 @@
+		}
+	}
+	return nil, lastErr
+}
+
+// Info returns the timestamping information. Caller MUST validate the timestamp
+// token info before consumption.
+func (t *SignedToken) Info() (*TSTInfo, error) {
+	var info TSTInfo
+	if _, err := asn1.Unmarshal(t.Content, &info); err != nil {
+		return nil, err
+	}
+	return &info, nil
+}
+
+// GetSigningCertificate gets the signing certificate identified by SignedToken
+// SignerInfo's SigningCertificateV2 attribute. If the IssuerSerial
+// field of signing certificate is missing, use signerInfo's sid instead.
+// The identified signing certificate MUST match the hash in
+// SigningCertificateV2.
+//
+// References: RFC 3161 2.4.1 & 2.4.2; RFC 5035 4
+func (t *SignedToken) GetSigningCertificate(signerInfo *cms.SignerInfo) (*x509.Certificate, error) {
+	var signingCertificateV2 signingCertificateV2
+	if err := signerInfo.SignedAttributes.TryGet(oid.SigningCertificateV2, &signingCertificateV2); err != nil {
+		return nil, fmt.Errorf("failed to get SigningCertificateV2 from signed attributes: %w", err)
+	}
+	// get candidate signing certificate
+	var candidateSigningCert *x509.Certificate
+	signed := (*cms.ParsedSignedData)(t)
+	if signingCertificateV2.Certificates[0].IssuerSerial.SerialNumber != nil {
+		// use IssuerSerial from SigningCertificateV2 signed attribute
+		if signingCertificateV2.Certificates[0].IssuerSerial.IssuerName.Name.Bytes == nil {
+			return nil, errors.New("issuer name is missing in IssuerSerial of SigningCertificateV2 attribute")
+		}
+		var issuer asn1.RawValue
+		if _, err := asn1.Unmarshal(signingCertificateV2.Certificates[0].IssuerSerial.IssuerName.Name.Bytes, &issuer); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal issuer name: %w", err)
+		}
+		ref := cms.IssuerAndSerialNumber{
+			Issuer:       issuer,
+			SerialNumber: signingCertificateV2.Certificates[0].IssuerSerial.SerialNumber,
+		}
+		candidateSigningCert = signed.GetCertificate(ref)
+	} else {
+		// use sid (unsigned) as IssuerSerial
+		candidateSigningCert = signed.GetCertificate(signerInfo.SignerIdentifier)
+	}
+	if candidateSigningCert == nil {
+		return nil, CertificateNotFoundError(errors.New("signing certificate not found in the timestamp token"))
+	}
+	// validate hash of candidate signing certificate
+	hash := crypto.SHA256 // default hash algorithm is id-sha256
+	var ok bool
+	if signingCertificateV2.Certificates[0].HashAlgorithm.Algorithm != nil {
+		// use hash algorithm from SigningCertificateV2 signed attribute
+		hash, ok = oid.ToHash(signingCertificateV2.Certificates[0].HashAlgorithm.Algorithm)
+		if !ok {
+			return nil, errors.New("unsupported certificate hash algorithm in SigningCertificateV2 attribute")
+		}
+	}
+	certHash, err := hashutil.ComputeHash(hash, candidateSigningCert.Raw)
+	if err != nil {
+		return nil, err
+	}
+	if !bytes.Equal(certHash, signingCertificateV2.Certificates[0].CertHash) {
```

Comment:
Why the signing certificate hash should match `CertHash`?

Reply:
This requirement is indeed very subtle, the order to find it is RFC 3161 -> [RFC 5816](https://datatracker.ietf.org/doc/html/rfc5816) -> [RFC 5035](https://datatracker.ietf.org/doc/html/rfc5035#section-3)